Add a max session time limit

boosh opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?
It allows users to specify the maximum duration of a transfer session sort of as requested in #985, making it easier to script rclone to e.g. backup during the night. The setting I've added introduces a duration flag. After rclone has been running for that amount of time, no new transfers will be initiated, but those already started will be able allowed to complete. It may be useful to implement a further flag for a hard limit that will terminate in-progress transfers once the session limit is reached, but in another PR.

Note - Since `--max-transfer` doesn't appear to have tests, neither does this feature. I have tested the non-server copy manually though - I've never used server mode so would need some pointers to test that.

#### Was the change discussed in an issue or in the forum before?
Sort of in #985

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

boosh commented:
OK thanks. Let me know if you want me to squash the commits.

boosh commented:
Ah it doesn't work if I move it to those places. I'll revert the change.

boosh commented:
`sync/run` is too high-level. I need to break during the transfer loop. Perhaps `moveOrCopy` didn't work because I tried syncing a directory? I expect that'd be the primary usecase since right now in-progress transfers aren't terminated once the session limit is reached. 

Is there anywhere else it needs adding to given it's mainly to help script backups (so several operations wouldn't be relevant)?

ncw commented:
Note that MaxTransfer does have tests here: fs/accounting/accounting_test.go :-)

I think that breaking the transfer in `Copy` is probably the wrong place.

What you want to do is [cancel the context](https://github.com/ncw/rclone/blob/a35aa1360e3e09d4973d52d995f7edaae803adb3/fs/sync/sync.go#L662) passed to the march routine.

I think using a deadline should work

```go
newCtx, cancel := context.WithDeadline(ctx, 100*time.Seconds) // config var goes here
// pass newCtx to march
// do the march
```

This will cause march routine to exit and cause the sync to come to a stop gracefully.

boosh commented:
OK I've changed to add a deadline to the context. I've added a test but it'll be quite brittle. Is there any way of throttling transfers during testing to e.g. simulate large files, slow connections etc? At the moment I've just created 10000 test files. The test works on my mac today but obviously as computers get faster it'll break one day... It'd be better to only create a handful of files with a function that just makes each transfer sleep for a while to make them safely reproducible.

boosh commented:
Hmm the appveyor test failed - perhaps because it's on Windows? On my Mac no `context.DeadlineExceeded` is ever thrown so I can't require an error in the tests. I assume it's something to do with the test filesystem abstraction. For now I won't add an assertion to the error returned from `Sync` unless there's a cross-platform solution. The test makes sure that not all files were transferred so that should be good enough.

ncw commented on fs/sync/sync.go:
```diff
@@ -657,6 +657,15 @@ func (s *syncCopyMove) run() error {
+		fs.Infof(s.fdst, "Transfer session deadline: %s", endTime.Format("2006/01/02 15:04:05"))
+		newContext, cancel := context.WithDeadline(s.ctx, endTime)
+		s.ctx = newContext
+		s.cancel = cancel
```
I don't think you want to overwrite the existing `s.ctx` or `s.cancel` as this context with deadline is just for March and its go-routines.

ncw commented on fs/sync/sync_test.go:
```diff
@@ -977,6 +978,35 @@ func TestSyncWithUpdateOlder(t *testing.T) {
+	maxSession, err := time.ParseDuration("2s")
+	require.NoError(t, err)
+
+	fs.Config.MaxSession = maxSession
```
You should put defer in here to reset the parameter back to the default

```
defer func(){
    fs.Config.MaxSession = 0 // reset back to default
}()
```

ncw commented on fs/sync/sync_test.go:
```diff
@@ -977,6 +978,35 @@ func TestSyncWithUpdateOlder(t *testing.T) {
+
+	fs.Config.MaxSession = maxSession
+
+	testFiles := make([]fstest.Item, 10000)
```
Making 10000 files will kill the integration tests with other backends for certain! (These tests are run against all the backends once per day).

How about just using a few filesa nd temporarily setting a bwlimit to 1b/s or something like that?

boosh replied:
They're still flaky on my Mac - I thought I'd got down to only needing 5 test files but sometimes it can get through a few hundred :-(

ncw commented:
> Hmm the appveyor test failed - perhaps because it's on Windows?

I think that is because you didn't reset the deadline in your tests see the `deadline exceeded` messages.

> On my Mac no `context.DeadlineExceeded` is ever thrown so I can't require an error in the tests.

You should find the error in `fs.Stats.GetLastError()`

This should work on all platforms

boosh commented:
Am I missing some kind of initialisation?

ncw commented:
I merged a lot of things which changes the stats interface which has caused the breakage.

If you want me to finish it off then I can - it is very nearly there.

If you want to have a go yourself, then rebase it onto master and change the uses of `fs.Config.Stats` to `accounting.GlobalStats()`

boosh commented:
@ncw It'd be great if you could finish it off please. I'm running into problems rebasing. Thanks.

ncw commented:
> @ncw It'd be great if you could finish it off please. I'm running into problems rebasing. Thanks.

No problems, will do :-)

boosh commented:
@88lex not yet.

ncw commented:
I'm still working on this, it turns out it is much harder than I first thought!

88lex commented:
> I'm still working on this, it turns out it is much harder than I first thought!
> @88lex not yet.

Completely understood. For all good things one must wait.

ncw commented:
All fixed now!  I'll merge it for 1.51
